Zoned occurrence using NodaTime

maknapp opened the pull request:
In response to #739 ([exact comment](https://github.com/orgs/ical-org/discussions/739#discussioncomment-14024476)), I started testing this out and got carried away with how much needed to change to make this work. This is only a draft proposal to change how evaluation works for version 6.

This changes the evaluation to use NodaTime and evaluates entirely in `ZonedDateTime`. Events are evaluated in their own time zone, or the specified time zone if floating, and all occurrences are returned in the specified time zone. Two tests fail because the expected data should be different when evaluated in a time zone. A few other tests fail because of NodaTime allowed values and formatting changes.

Notes:
* Evaluation uses only NodaTime, so most of the math operations can be removed from `CalDateTime`.
* I only got benchmarks to build, but did not run any.
* For backward compatibility, I think `EvaluationPeriod` has all the information to produce the same DtStart, DtEnd/Duration, and time zone values as before, but I only did DtStart for passing tests.
* I left many documentation comments alone, so many probably do not match anymore.

maknapp committed:
Use BCL HashCode

maknapp committed:
Evaluate recurrence with nodatime

maknapp committed:
Fix date parsing in a test

maknapp committed:
Evaluate calendar in a time zone always

This evaluates events in the calendar for a specific time zone.
Events with a time zone are evaluated in that event's time zone and then
converted to the requested time zone. Floating events are evaluated
in the request time zone. This is essentially the same behavior as before,
except the results are actual point-in-time values.

A key evaluation difference is that during backward daylight saving,
the last time offset can be output instead of only the first. Before,
an hourly repeating event would skip the second hour.

Fixes #681, #875

maknapp committed:
Remove invalid comparison operators

The only valid comparison for two CalDateTime values is equality.
The greater and less than operators cannot always return a valid
result. This makes implementing IComparable also invalid.
"20250815" != "20250815T000000" but these operators also did not
say they were less or greater either.

maknapp committed:
Remove copying of CalDateTime

CalDateTime is immutable, so there is no need to copy.

maknapp committed:
Add backward daylight saving tests

maknapp committed:
Change benchmarks to fit zoned API

maknapp committed:
Store CalDateTime values with NodaTime

maknapp committed:
Reduce evaluation allocation

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -23,12 +24,15 @@ namespace Ical.Net.DataTypes;
 /// </remarks>
 /// </summary>
-public sealed class CalDateTime : IComparable<CalDateTime>, IFormattable
+public sealed class CalDateTime : IFormattable
```
Finally this should become a small, dump iCalendar wall time representation. Wall time arithmetic could go into extension methods.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -373,67 +316,144 @@ public DateTime Value
+    /// Gets the day.
     /// </summary>
-    public int Year => Value.Year;
+    public int Day => _localDate.Day;
```
Do the wrapped Day, Month, Year, Hour etc. properties bring a benefit to users?
Exposing `DateTime` could be sufficient?

maknapp replied:
They are useful to have and simple enough to keep. They are used significantly in tests. Anyone using `CalDateTime` directly would probably prefer `if (calDateTime.Year == 2025)` over `if (calDateTime.Date.Year == 2025)`. For comparison, `NodaTime` types provide these wrapped properties everywhere.

The time values being nullable vs defaulting to 0 needs some thought though. Nullable is more accurate. Defaulting to 0 matches the values of `ToLocalDateTime()` and `Value`.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -373,67 +316,144 @@ public DateTime Value
-    public int DayOfYear => Value.DayOfYear;
+    public LocalTime? Time => _localTime;
+
+#if NET6_0_OR_GREATER
```
The pragma is okay for the CTOR, but should public properties be different across target frameworks?

maknapp replied:
I changed them to methods `ToDateOnly()` and `ToTimeOnly()` to match `NodaTime`. Is there a specific argument against this?

axunonb replied:
Yes, it would be good to have the same public API for all target frameworks. The comment is about the #if NET6_0_OR_GREATER pragma influencing the public API.
As the pkg `Portable.System.DateTimeOnly` is removed and internally replaced by NodaTime types, what would be the purpose for `TimeOnly` / `DateOnly`? So I'd vote for dropping `TimeOnly` and `DateOnly` altogether.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -373,67 +316,144 @@ public DateTime Value
+    {
+        if (_tzId is null)
+        {
+            return ToLocalDateTime().InUtc();
+        }
+        else
+        {
+            return DateUtil.GetZone(_tzId).AtLeniently(ToLocalDateTime());
+        }
+    }
+
+    public ZonedDateTime ToZonedDateTime(DateTimeZone timeZone)
+    {
+        if (_tzId is null)
+        {
+            return ToLocalDateTime().InZoneLeniently(timeZone);
+        }
+        else
+        {
+            return DateUtil.GetZone(_tzId)
+                .AtLeniently(ToLocalDateTime())
+                .WithZone(timeZone);
+        }
+    }
+
+    public ZonedDateTime ToZonedDateTime(string zoneId)
+    {
+        return ToZonedDateTime(DateUtil.GetZone(zoneId));
+    }
+
+    public ZonedDateTime AsZonedOrDefault(DateTimeZone timeZone)
+    {
+        if (_tzId is null)
+        {
+            return ToLocalDateTime().InZoneLeniently(timeZone);
+        }
+        else
+        {
+            return DateUtil.GetZone(_tzId).AtLeniently(ToLocalDateTime());
+        }
```
For evaluation we could use an internal class EvaluationDateTime or alike. NodaTime types cannot cover all iCalendar evaluation rules.

axunonb commented on Ical.Net/DataTypes/Duration.cs:
```diff
@@ -45,6 +45,43 @@ public Duration(int? weeks = null, int? days = null, int? hours = null, int? min
         Seconds = seconds;
     }
 
+    public NodaTime.Period ToPeriod()
```
Change members returning NodaTime.Period to internal?

axunonb commented on Ical.Net/DataTypes/Duration.cs:
```diff
@@ -170,18 +207,6 @@ internal static Duration FromTimeSpan(TimeSpan t)
-    /// To convert a duration to a <see cref="TimeSpan"/> while considering the days and weeks as nominal durations,
-    /// use <see cref="ToTimeSpan"/>.
-    /// </remarks>
-    public TimeSpan ToTimeSpan(CalDateTime start)
```
Yep, remove evaluation

axunonb commented on Ical.Net/DataTypes/FreeBusyEntry.cs:
```diff
@@ -20,6 +23,11 @@ public FreeBusyEntry(Period period, FreeBusyStatus status)
+
+        if (EndTime is null && Duration is null)
+        {
+            throw new ArgumentException("Period must have a duration", nameof(period));
```
"Period must either have a Duration or an EndTime"?
Or should we create a FreeBusyEntry from EvaluationPeriod?

maknapp replied:
I fixed this error message. `FreeBusyEntry` is `Period` with a FBTYPE (`FreeBusyStatus`) and the requirement that datetimes MUST be UTC. Nothing verifies that dates are UTC here when it probably should.

axunonb commented on Ical.Net/DataTypes/Occurrence.cs:
```diff
@@ -5,46 +5,106 @@
-    public Period Period { get; set; }
-    public IRecurrable Source { get; set; }
+    public IRecurrable Source { get; private set; }
+    public ZonedDateTime Start { get; private set; }
```
Expose ZonedDateTime for Start and End, or the wall time (CalDateTime) - do we need both?
Make other members internal?

maknapp replied:
`NodaTime.ZonedDateTime` has all the data involved and is what is used while evaluating. Other types like `NodaTime.Instant` or `NodaTime.OffsetDateTime` can be used, but time zone would be lost. Since time zone is known already (because `GetOccurrences()` requires it), it is not unreasonable to drop the time zone and assume that the user knows that ALL occurrences are in the given time zone. `CalDateTime` can represent an exact time, but it is ambiguous so not very helpful. IF we want to hide `NodaTime` completely, using `DateTimeOffset` for `Start` and `End` would be the most accurate option.

axunonb commented on Ical.Net/DataTypes/Period.cs:
```diff
@@ -78,8 +82,23 @@ public Period(CalDateTime start, CalDateTime? end = null)
 
-        if (end != null && end.LessThan(start))
-            throw new ArgumentException($"End time ({end}) must be greater than start time ({start}).", nameof(end));
+        if (end != null)
```
Make `Period` as dump as `CalDateTime` - a pure iCalendar duration without evaluation?

axunonb commented on Ical.Net/Evaluation/EvaluationPeriod.cs:
```diff
@@ -0,0 +1,52 @@
+    public ZonedDateTime? End { get; private set; }
+
+    public EvaluationPeriod(ZonedDateTime start, ZonedDateTime? end = null)
+    {
```
Move logic that currently exist in `Period`, e.g. start >= end etc.?

axunonb commented on Ical.Net/Evaluation/EvaluationPeriod.cs:
```diff
@@ -0,0 +1,52 @@
+using NodaTime;
+
+namespace Ical.Net.Evaluation;
+public class EvaluationPeriod : IComparable<EvaluationPeriod>
```
Could all Evaluation classes become internal. Would increase flexibility for future changes without causing a breaking change?

maknapp replied:
I'm not sure. Some people might be using `RecurrencePatternEvaluator` directly for expanding RRULEs without using calendar features - similar to using rrule.js. Forcing the use of `Calendar` would have more allocations and probably be slightly slower.

axunonb commented on Ical.Net/NodaTimeExtensions.cs:
```diff
@@ -0,0 +1,48 @@
+using NodaTime.TimeZones;
+
+namespace Ical.Net;
+internal static class NodaTimeExtensions
```
Nice!

axunonb commented on Ical.Net/Serialization/DataTypes/DateTimeSerializer.cs:
```diff
@@ -86,18 +87,18 @@ public DateTimeSerializer(SerializationContext ctx) : base(ctx) { }
 
-        var datePart = new DateOnly(); // Initialize. At this point, we know that the date part is present
-        TimeOnly? timePart = null;
+        var datePart = new LocalDate(); // Initialize. At this point, we know that the date part is present
```
Yep, pure wall time

axunonb reviewed (commented):
I created a branch `version/6.0` where this PR could get merged.
Unit test that currently fail should be modified.
Considering the size of this PR, which next step would you prefer?

maknapp committed:
Fix CalDateTime format test

maknapp committed:
Validate CalDateTime year

CalDateTime was validating year using DateTime limits. NodaTime
supports more than just positive years, so verify that the year
is positive to match DateTime behavior.

maknapp committed:
Change out of range test

This test had an effective end that exceeded the valid date
range but only actually held a duration, so there was no
exception. The new occurrence format now produces an end
date always, which now exceeds the valid date range.

maknapp committed:
Change time zone test

This was essentially testing in the system time zone before.
Now that evaluation requires a time zone, this required
adjusting the local time to fit the default testing time zone
of US Eastern.

maknapp committed:
Removed warning

It makes sense to be a nominal duration in this case,
even if the spec does not say so specifically.

maknapp committed:
Validate EvaluationPeriod end

maknapp committed:
Fix freebusy validation message

maknapp committed:
Simplify CalDateTime LocalDateTime conversion

maknapp committed:
Change dateonly timeonly to methods

maknapp committed:
Remove unused DateUtil methods

maknapp committed:
Fix some sonar issues

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -215,17 +228,12 @@ public CalDateTime(string value, string? tzId = null)
-    {
-        _dateOnly = dateOnly;
-        _timeOnly = TruncateTimeToSeconds(timeOnly);
+#if NET6_0_OR_GREATER
```
Would be good to have the same public API for all target frameworks.
As the pkg `Portable.System.DateTimeOnly` and internally replace by NodaTime types, is there a reason to keep support for `TimeOnly` / `DateOnly`?

maknapp replied:
Supporting conversion to available BCL types is more important to me than keeping the public API the same. This is just wrapping the same conditional API that NodaTime has.

It allows direct access to the underlying CalDateTime data for people that do not want to use NodaTime. If someone did not want to use NodaTime, I would suggest using DateOnly and TimeOnly over DateTime if possible. This is even more important if we cut out all NodaTime from the public API.

axunonb replied:
> This is even more important if we cut out all NodaTime from the public API.

Argee in this aspect. And sure, `DateTime` is not sufficient in case of date-only.
So let's leave it as it is for now.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -23,12 +25,15 @@ namespace Ical.Net.DataTypes;
 /// </remarks>
 /// </summary>
-public sealed class CalDateTime : IComparable<CalDateTime>, IFormattable
+public sealed class CalDateTime : IFormattable, IEquatable<CalDateTime>
```
This would happen **after** merging this PR?

`CalDateTime` (The "Purified" Class)
This class would be stripped of all logic and become a pure data object.

  * **It would ONLY contain:**
      * `public DateTime Value { get; }` (The raw `System.DateTime` read from the file)
      * `public string Tzid { get; }` (The raw `TZID` string, e.g., "America/New\_York")
      * `public bool IsUtc { get; }`
      * `public bool IsDateOnly { get; }`
  * **It would NOT contain all the evaluation logic:**
      * Any `NodaTime` properties (`ZonedDateTime`, `DateTimeZone`).
      * All math methods (`AddHours`, `AddDays`, `Subtract`).
      * All comparison methods (`CompareTo`, `Equals`, `GetHashCode` would just compare the raw data values).
      * All conversion methods (`ToUniversalTime`, `ToTimeZone`).

See also #849 - arithmetics with CalDateTime are painful...

maknapp replied:
Yes. I'm still voting for keeping some of the NodaTime properties on CalDateTime. Removing the math methods definitely. _Maybe_ keep some conversion methods. I think each group could be a discussed in separate PRs.

axunonb replied:
That's the dilemma. Doing so is convenient, but polluting the evaluation layer.
NodaTime properties are okay for internal use, If we absolutely need conversion methods they should go into extension methods for `CalDateTime`, keeping the class as lean and dumb as necessary for an iCalendar data type.

maknapp committed:
Change tests to fit zoned API

The few tests still failing were left alone on purpose. With some tests,
the test data needs to be changed because it does not make sense now that
all occurrences are in a specific time zone.

For FREEBUSY tests, that was mostly ignored because I am not sure how it
should be used.

axunonb reviewed (approved):
Thanks, ready for next steps
